### ISSUE TITLE
Sanitizing function false positive for do_shortcode() in HTML context

### ISSUE THREAD

**Issue by GaryJones** [WordPress-Coding-Standards/WordPress-Coding-Standards#167]
`<li class="menu-item live-person"><?php echo do_shortcode( '[liveperson]' ); ?></li>`

This flags up as "Expected a sanitizing function (see Codex for 'Data Validation'), but instead saw 'do_shortcode'", and I think this is incorrect.

In addition, note that this is expecting a sanitization (which should only happen on input) and not an escaping function (which happens on output), so that appears to be incorrect or badly worded too.

**Comment by rraub** [comment]
:+1: 
I've run into this as well, here is the obvious work around:
```php
<?php
 // @codingStandardsIgnoreStart
 echo do_shortcode( '[liveperson]' ); 
// @codingStandardsIgnoreEnd
?>
```

**Comment by lkraav** [comment]
Opened nearly a year ago. Is there a definitive solution here of some sort? I guess ignoring standards where situation warrants it is also a valid solution, feels a bit noisy though.

**Comment by JDGrimes** [comment]
I think this is really the correct behavior. @GaryJones might agree after #159, #282. While it may seem like the onus to properly escape output should be on the shortcode function, the principle of "late escaping" is that the onus is *always* at the location where `echo` is called. 

There are several workarounds in addition to [the one given by @rraub](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/issues/167#issuecomment-73079613).

You can add a comment telling the XSS sniff to ignore this:

```php
echo do_shortcode( '[liveperson]' ); // XSS OK
```

You can add `do_shortcode()` to the list of auto-escaped functions in your XML config file:

```xml
	<rule ref="WordPress.XSS.EscapeOutput">
		<properties>
			<property name="customAutoEscapedFunctions" value="do_shortcode" type="array" />
		</properties>
	</rule>
```

You can wrap the call in `wp_kses()` (this is probably the best way):

```php
echo wp_kses( do_shortcode( '[liveperson]' ), 'post' );
```

However, if your case is exactly like this, [you probably shouldn't be using `do_shortcode()`](https://kovshenin.com/2013/dont-do_shortcode/) anyway.

**Comment by GaryJones** [comment]
Under what context should the escaping happen? It's not an attribute, or JS, or a URL, and in the case of my example above, it happens to output HTML (for a live chat feature), so escaping HTML would stop it displaying as HTML. There's no existing context for me to escape with.

**Comment by JDGrimes** [comment]
> Under what context should the escaping happen? It's not an attribute, or JS, or a URL, and in the case of my example above, it happens to output HTML (for a live chat feature), so escaping HTML would stop it displaying as HTML. There's no existing context for me to escape with.

For HTML, use `wp_kses()`, if possible. 

But really I wouldn't do it this way at all. I'd remove the use of `do_shortcode()` completely, and just call the shortcode callback function directly. Like, `liveperson_shortcode()`. Then, if that function is under your control, you can add it to the `customAutoEscapedFunctions` via XML config as shown above.

Otherwise, if you just use `do_shortcode()`, you have no way of knowing what code will be generating the shortcode content. Something might be calling `add_shortcode()` with a different function than you expect. There is no certainty of the output when using `do_shortcode()`, so we can't add it to the list of auto-escaped functions. It doesn't match this (from https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/pull/282#issuecomment-66634468):

>Any non-pluggable function that returns a known limited set of safe static values not directly based on the input arguments

I realize that all of this might seem overly-cautious. And maybe it is, I'm just trying to follow the reasoning of the VIP team on late escaping. I don't think they'd consider it a false positive (though I'm willing to be corrected). The result *should* be escaped, but there is no guarantee that it will be.

I think the real issue here is that `do_shortcode()` is being used at all. You should really almost never need to use that function, except occasionally attaching it to a filter. Or, if you must use it, you can escape or ignore it on a case-by-case basis. So, I think that using `do_shortcode()` like this isn't best practice to start with.

**Comment by GaryJones** [comment]
> I'd remove the use of do_shortcode() completely, and just call the shortcode callback function directly. Like, liveperson_shortcode(). Then, if that function is under your control, you can add it to the customAutoEscapedFunctions via XML config as shown above.
> 
> Otherwise, if you just use do_shortcode(), you have no way of knowing what code will be generating the shortcode content. Something might be calling add_shortcode() with a different function than you expect.

Since this comes from a plugin, I think that's kind of the point - themes can call `do_shortcode()`, but the plugin author can update the function name or migrate it to a class method as needed without breaking back compat.

However, accepting and closing this as wontfix.

**Comment by JDGrimes** [comment]
> Since this comes from a plugin, I think that's kind of the point - themes can call do_shortcode(), but the plugin author can update the function name or migrate it to a class method as needed without breaking back compat.

That is why https://core.trac.wordpress.org/ticket/25435.
